### PR TITLE
BUG: fix support for gufuncs with more than one core dimension

### LINKED ISSUE #16116: BUG: `MaskedNDArray.__array_ufunc__` breaks for `erfa.rxp`
I stumbled upon this while working on #16070. This looks like a bug in `MaskedNDArray.__array_ufunc__`, but maybe it's really a bug in pyerfa ?
I'm opening this report as soon as I managed to isolate the problem, but I suspect the issue is more general than it looks at first glance.

```python
import erfa
import numpy as np
from astropy.utils.masked import Masked
import pytest


@pytest.mark.parametrize(
    "m, v",
    [
        (np.eye(3), np.arange(3)), # this works
        (Masked(np.eye(3)), Masked(np.arange(3))), # this doesn't
    ],
)
def test_reg(m, v):
    erfa.rxp(m, v)
```

relevant pytest output
```
___________________________________________ test_reg[m1-v1] ___________________________________________

m = MaskedNDArray([[1., 0., 0.],
               [0., 1., 0.],
               [0., 0., 1.]])
v = MaskedNDArray([0, 1, 2])

    @pytest.mark.parametrize(
        "m, v",
        [
            (np.eye(3), np.arange(3)),
            (Masked(np.eye(3)), Masked(np.arange(3))),
        ],
    )
    def test_reg(m, v):
>       erfa.rxp(m, v)

g.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/clm/.pyenv/versions/astropy.dev/lib/python3.12/site-packages/erfa/core.py:19906: in rxp
    rp = ufunc.rxp(r, p)
../../astropy/utils/masked/core.py:872: in __array_ufunc__
    return self._masked_result(result, mask, out)
../../astropy/utils/masked/core.py:952: in _masked_result
    return Masked(result, mask)
../../astropy/utils/masked/core.py:77: in __new__
    return cls._get_masked_instance(*args, **kwargs)
../../astropy/utils/masked/core.py:125: in _get_masked_instance
    return masked_cls.from_unmasked(data, mask, copy)
../../astropy/utils/masked/core.py:518: in from_unmasked
    self._set_mask(mask, copy=copy)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MaskedNDArray([0., 1., 2.]), mask = array([[False],
       [False],
       [False]])
copy = False

    def _set_mask(self, mask, copy=False):
        self_dtype = getattr(self, "dtype", None)
        mask_dtype = (
            np.ma.make_mask_descr(self_dtype)
            if self_dtype and self_dtype.names
            else np.dtype("?")
        )
        ma = np.asanyarray(mask, dtype=mask_dtype)
        if ma.shape != self.shape:
            # This will fail (correctly) if not broadcastable.
            self._mask = np.empty(self.shape, dtype=mask_dtype)
>           self._mask[...] = ma
E           ValueError: could not broadcast input array from shape (3,1) into shape (3,)

../../astropy/utils/masked/core.py:238: ValueError
======================================= short test summary info =======================================
FAILED g.py::test_reg[m1-v1] - ValueError: could not broadcast input array from shape (3,1) into shape (3,)
```

### PR DESCRIPTION
**Pull request by mhvk**
In particular, `erfa.rxp` and `erfa.rxr`.

This just fixes the bug, and is slightly weird in its treatment of `axis`. I have a follow-up PR in the works which just adds proper support for `axes` - but that would be less suitable for backporting.

Fixes #16116

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Review by neutrinoceros — APPROVED** [review]
I find it pretty elegant for a hack :-)
More importantly, it does fix the issue in a back portable way, so that's more than enough for me. Thank you very much for going at it so quickly !

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
Actually, now I'm seeing a new error in `MaskedArray.__array_ufunc__` in #16070 (which this PR is supposed to unblock)  https://github.com/astropy/astropy/actions/runs/8076412008/job/22064849015?pr=16070

I don't know yet if that's a new bug in this PR or if it's also present on main (will inspect). I just wanted to hold off this one until we know more. I'll update my review status as soon as possible !

**Review by neutrinoceros — APPROVED** [review]
Okay, so what I'm seeing on #16070 is indeed another, unrelated (but adjacent) bug, see https://github.com/astropy/astropy/issues/16123

I don't know that it should be fixed alongside #16116 (maybe so ?), in any case, I think it's okay to merge this one as is. Thanks again !

**Comment by mhvk** [comment]
@pllim - this was reviewed by @neutrinoceros, which I think should suffice. Could you merge for us?

**Review by pllim — APPROVED** [review]
Approving by proxy. Thanks!

(I don't grok this but I trust you...)

**Pull request merged**
